### PR TITLE
Basic sampling

### PR DESCRIPTION
**Pull request by yannikschaelte**
Adds:

* `pypesto.sample` function
* `pypesto.sample.Sampler` and `pypesto.sample.InternalSampler` (for parallel tempering) interface base classes
* `pypesto.sample.McmcPtResult` result object
* Own Metropolis, Adaptive Metropolis sampler implementations
* Own Parallel Tempering, Adaptive Parallel Tempering sampler implementations
* Basic visualization routines
* tests for the new samplers
* usage example notebook

### CODE REVIEW

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/visualize/sampling.py  7
- test/test_sampling.py  5
         

Clones added
============
- pypesto/visualize/sampling.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5328569&bid=17441185)

File: pypesto/sample/sample.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        result: Result = None
+) -> Result:
+    """
+    This is the main function to call to do parameter sampling.
+
+    Parameters
+    ----------
+    problem:
+        The problem to be solved. If None is provided, a
+        :class:`pypesto.Pymc3Sampler` is used.
```
Comment (FFroehlich) [inline pypesto/sample/sample.py:26]:
```suggestion
        :class:`pypesto.Sampler` is used.
```

File: pypesto/profile/result.py [hunk 2]
```diff
@@ -5,7 +5,7 @@ class ProfilerResult(dict):
     """
     The result of a profiler run. The standardized return return value from
     pypesto.profile, which can either be initialized from an OptimizerResult
-    or from an existing ProfilerResult (in order to extent the compputation).
+    or from an existing ProfilerResult (in order to extend the compputation).
```
Comment (elbaraim) [inline pypesto/profile/result.py:8]: typo `compputation` :see_no_evil:

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/visualize/sampling.py  7
- test/test_sampling.py  5
         

Clones added
============
- pypesto/visualize/sampling.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5328569&bid=17441185)

File: pypesto/sample/adaptive_parallel_tempering.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+import numpy as np
+
+from .parallel_tempering import ParallelTemperingSampler
+
+
+class AdaptiveParallelTemperingSampler(ParallelTemperingSampler):
+    """Parallel tempering sampler with adaptive temperature adaptation."""
+
+    @classmethod
+    def default_options(cls) -> Dict:
```
Comment (elbaraim) [inline pypesto/sample/adaptive_parallel_tempering.py:11]: what about adding default values for (as in matlab PESTO)
```
'exp_temperature': 4
'max_temp': 50000
```

File: pypesto/sample/sample.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        A result with filled in sample_options part.
+    """
+    # prepare result object
+    if result is None:
+        result = Result(problem)
+
+    # try to find initial parameters
+    if x0 is None:
+        result.optimize_result.sort()
+        xs = result.optimize_result.get_for_key('x')
```
Comment (elbaraim) [inline pypesto/sample/sample.py:51]: how is this handled for parallel tempering?

Comment (yannikschaelte) [reply pypesto/sample/sample.py:51]: currently, all chains are initialized to the same starting point. of course, one could apply a clustering analysis and start from different starting points at some point, not sure about that ...

Comment (elbaraim) [reply pypesto/sample/sample.py:51]: I think it does not make sense to start all chains at the same point, then you lose the advantage of PT somehow.... How I usually do it is to look at the waterfall plot from the optimization and there I choose the different theta vectors for the initialization, i.e., different local optima.

Comment (yannikschaelte) [reply pypesto/sample/sample.py:51]: Sounds good. Not sure if there are theoretical results on starting PT chains from different points. I would assume that chains at high temperatures diverge fast anyway, but yes, mixing will be better if the starting points are different. To automatize, a clustering analysis will be necessary, which partly exists already in the visualization functions.

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/visualize/sampling.py  7
- test/test_sampling.py  5
         

Clones added
============
- pypesto/visualize/sampling.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5328569&bid=17441185)

File: pypesto/sample/sample.py (hunk 4 above)
Comment (yannikschaelte) [reply pypesto/sample/sample.py:53]: I added now the possibility to manually define multiple starts. for automatic schemes I suggest to do that separately, in particular once priors are defined.

File: pypesto/sample/adaptive_metropolis.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+        iteration.
+    cov:
+        The estimated covariance matrix of the sample, that was calculated in
+        the previous iteration.
+    x_new:
+        Most recent sample.
+    n_cur_sample:
+        Current number of samples.
+    decay_constant:
+        Adaption decay, in (0, 1). Higher values result in faster decays, such
```
Comment (dilpath) [inline pypesto/sample/adaptive_metropolis.py:112]: [0, 1] in default_options method

File: pypesto/sample/adaptive_metropolis.py [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+    mean = (1 - update_rate) * mean + update_rate * x_new
+
+    dx = x_new - mean
+    cov = (1 - update_rate) * cov + \
+        update_rate * dx.reshape((-1, 1)) @ dx.reshape((1, -1))
+
+    return mean, cov
+
+
+def regularize_covariance(cov: np.ndarray, reg_factor: float):
```
Comment (dilpath) [inline pypesto/sample/adaptive_metropolis.py:132]: return -> np.ndarray

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/visualize/sampling.py  7
- test/test_sampling.py  5
         

Clones added
============
- pypesto/visualize/sampling.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5328569&bid=17441185)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/visualize/sampling.py  7
- test/test_sampling.py  5
         

Clones added
============
- pypesto/visualize/sampling.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5328569&bid=17441185)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- test/test_sampling.py  5
- pypesto/visualize/sampling.py  7
         

Clones added
============
- pypesto/visualize/sampling.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5328569&bid=17441185)

**Pull request merged**